SPIFlashLogger.class.nut

```diff
@@ -76,53 +114,185 @@ class SPIFlashLogger {
             }
 
             // Now write the data
-            _write(obj, _at_sec, _at_pos, obj_pos, sec_remaining);
-            _map[_at_sec] = SPIFLASHLOGGER_SECTOR_DIRTY;
+            _write(obj, _atSec, _atPos, objPos, secRemaining);
+            _map[_atSec] = SPIFLASHLOGGER_SECTOR_DIRTY;
 
             // Update the positions
-            obj_pos += sec_remaining;
-            obj_remaining -= sec_remaining;
-            _at_pos += sec_remaining;
-            if (_at_pos >= SPIFLASHLOGGER_SECTOR_SIZE) {
-                _at_sec = (_at_sec + 1) % _sectors;
-                _at_pos = SPIFLASHLOGGER_SECTOR_META_SIZE;
+            objPos += secRemaining;
+            objRemaining -= secRemaining;
+            _atPos += secRemaining;
+            if (_atPos >= SPIFLASHLOGGER_SECTOR_SIZE) {
+                _atSec = (_atSec + 1) % _sectors;
+                _atPos = SPIFLASHLOGGER_SECTOR_METADATA_SIZE;
             }
-        } while (obj_remaining > 0);
+        } while (objRemaining > 0);
 
         _disable();
     }
 
+    // Reads objects from the logger synchronously,
+    // returning a single log object for the specified *index*.
+    //
+    // Returns:
+    // Deserialized object or null
+    //    - the most recent object when `index == -1`,
+    //    - the oldest object when `index == 1`,
+    //    - *null* when the value of *index* is greater than the number of logs,
+    //    - throws an exception when `index == 0`.
+    //
+    function readSync(index = 0) {
+        // Unexpected index value
+        if (index == 0)
+            throw "Invalid argument";
+
+        // identify first sector to read
+        local seekTo = math.abs(index);
+        local count = 0;
+        local i = 0;
+
+        // _atSec - indicates the current write position, therefore
+        // for the index > 0 it is necessary to step forward (skip current sector)
+        // to find first not empty sector
+        // For the index < 0, it is possible to start couting from the current sector
+        while (i < _sectors) {
+            // convert sector index `i`, ordered by recency, to physical `sector`, ordered by position on disk
+            local sector;
+            if (index > 0) {
+                sector = (_atSec + i + 1) % _sectors;
+            } else {
+                sector = (_atSec - i + _sectors) % _sectors;
+            }
+
+            ++i;
+
+            // Read all objects start codes for current sector
+            // and write them into the blob
+            local objectsStartCodes = _getObjectsStartCodesForSector(sector);
+
+            // If there is no start codes in the current sector
+            // then switch to the next sector
+            if (objectsStartCodes == null || objectsStartCodes.len() == 0)
+                continue;
+
+            // negative step, go backwards
+            if (index < 0)
+                objectsStartCodes.seek(-2, 'e');
+
+            // Check that the number of start codes is enough
```

Comment:
It would probably be good to have a tests that read last objects in sectors in both directions. Do we have ones?

Reply:
Ok, I'll add such test

Reply:
Thank you!